### PR TITLE
feat(machine): implement keep instance on new machine domain

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch adds the functionality of keep-instance to the new machine domain.

Since the two methods `SetKeepInstance()` and `KeepInstance()` were only used in two facades, the call to the legacy state has been already deleted and therefore the entire functionality lives in Dqlite without the need for double writing.

_Note: Since this feature is so simple and self-contained, the whole functionality (from ddl changes to facades wiring) has been implemented in this only patch, without the need for any temporary double writes._

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

The keep-instance functionality must continue working as usual, no change expected even though the full functionality lives in dqlite now.

Example on lxd:

```
juju bootstrap localhost c
juju add-model m
juju add-machine
lxc ls
+---------------+---------+-----------------------+------+-----------+-----------+
|     NAME      |  STATE  |         IPV4          | IPV6 |   TYPE    | SNAPSHOTS |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-365cdd-0 | RUNNING | 10.170.236.168 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-0 | RUNNING | 10.170.236.132 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+

juju remove-machine 0 --keep-instance
WARNING This command will perform the following actions:
will remove machine 0 (but retaining cloud instance)

Continue [y/N]? y
lxc ls
+---------------+---------+-----------------------+------+-----------+-----------+
|     NAME      |  STATE  |         IPV4          | IPV6 |   TYPE    | SNAPSHOTS |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-365cdd-0 | RUNNING | 10.170.236.168 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-0 | RUNNING | 10.170.236.132 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
```
Then we add another machine and when we delete it without --keep-instance, we expect it to be removed from lxd.
```
juju add-machine
lxc ls
+---------------+---------+-----------------------+------+-----------+-----------+
|     NAME      |  STATE  |         IPV4          | IPV6 |   TYPE    | SNAPSHOTS |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-365cdd-0 | RUNNING | 10.170.236.168 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-0 | RUNNING | 10.170.236.132 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-1 | RUNNING | 10.170.236.27 (eth0)  |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
juju remove-machine 1
WARNING This command will perform the following actions:
will remove machine 1

Continue [y/N]? y
lxc ls
+---------------+---------+-----------------------+------+-----------+-----------+
|     NAME      |  STATE  |         IPV4          | IPV6 |   TYPE    | SNAPSHOTS |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-365cdd-0 | RUNNING | 10.170.236.168 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-0 | RUNNING | 10.170.236.132 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
```

## Links

**Jira card:** JUJU-6702

### CODE REVIEW

File: domain/machine/state/state.go [hunk 1]
```diff
@@ -776,3 +777,78 @@ func (st *State) GetMachineUUID(ctx context.Context, name machine.Name) (string,
...
+		affected, err := outcome.Result().RowsAffected()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if affected != 1 {
+			return machineerrors.MachineNotFound
+		}
+		if err != nil {
+			return fmt.Errorf("setting keep instance for machine %q: %w", mName, err)
+		}
```
Comment (SimonRichardson) [inline domain/machine/state/state.go:851]: I'd rather we split this into two queries, one to get the machine and then one to update. We're then not attempting to write if the machine doesn't exist.

Comment (nvinuesa) [reply domain/machine/state/state.go]: Yeah, nice catch. See 7699e5140241109e1fa299cb575ebf398c6fab75

**Review by SimonRichardson — APPROVED** [review]
Once the triggers/mocks have been generated, I'm happy for this to land.

**Pull request merged**